### ISSUE TITLE
Feature request: allow using QtAV as static/shared library *without* installation into Qt folder

### ISSUE THREAD

**Issue by ElderOrb** [nejatafshar/QtAV#3]
I've asked for this functionality in official QtAV repo but got no reply. 

My goal is to keep build steps as simple as possible and to be able to use several versions of QtAV in parallel (for experiments). Is there any fundamental issues disallowing to *just* include QtAV pri file into application's pro file, or QtAV pro file into subtree project and that's it?

**Comment by nejatafshar** [comment]
@ElderOrb Installing QtAV to to Qt folder is a mechanism to simplifiy using it. But  nothing stops you from linking  to a QtAV build in your subproject. All you have to do is define `LIBS` and `INCLUDEPATH` to the desired build and include path of QtAV lib instead of `QT += av`.

**Comment by ElderOrb** [comment]
But in this case I have to use pre-built QtAV, right? I hoped to build QtAV together with project (to ensure I'm always using the latest and greatest). And for this would be helpful to have QtAV.pri file.

**Comment by ElderOrb** [comment]
Also, not sure what is wrong, but on attempt to build QtAV under windows (from QtCreator) I'm getting this: 

Checking for avutil... no
Project ERROR: avutil is required, but compiler can not find it

I've added 

message('INCLUDPATH: ' $$INCLUDEPATH)
message('LIBS: ' $$LIBS)

 and reviewed output - QtAV is using ffmpeg from 'ffmpegBin'. What I'm missing?

**Comment by nejatafshar** [comment]
@ElderOrb You can have QtAV as a subproject in a subdirs project. This way it is built in a subdirectory in your build path and you can just link to relative built libs.
About the error you mentioned, may be you have not retrieved FFmpeg files properly. The FFmpeg files in `ffmpegBin` directory are tracked as lfs. Are you sure you have git lfs installed?

**Comment by ElderOrb** [comment]
Okay, QtAV as a subproject should also work (if it will work). As for the ffmpeg files - I didn't retrieve anything additionally. I just thought the reason for having dlls, libs and headers in 'ffmpegBin' folder - is because they are used during QtAV build by default. As for the git lfs - I've personally never used it, but can confirm that it is installed. 

So what should be right steps for building your fork for windows? Besides git clone && opening in QtCreator I mean..

**Comment by nejatafshar** [comment]
@ElderOrb The ffmpeg files in ffmpegBin dir are used by default. They are tracked in git repository as lfs files. Try to initialize git lfs by running `git lfs install` command once in git shell and clone the repository again.

**Comment by ElderOrb** [comment]
Hmm, now it builds, thanks a lot! Although I'm not quite sure whether this is because of 'git lfs install' and re-cloning, or because I'm on different environment. I've even compared cloned version before 'git lfs install' with the one cloned after 'git lfs install' - seems pretty the same, with the only difference inside .git folder. 

Anyway, as the next step will try to integrate it into the subdir project.

**Comment by nejatafshar** [comment]
@ElderOrb if you clone repository without initializing git lfs once, you would see all lfs files there but with small sizes like afew kilo bytes.

**Comment by ElderOrb** [comment]
Unfortunately adding QtAV.pro as a subdir doesn't work (at least on Windows).

**Comment by ElderOrb** [comment]
No matter what I'm doing - output is 'QtAVd.dll' (should be QtAVd1.dll as I understand) - as the result Widgets and QML integration fails to build.

**Comment by ElderOrb** [comment]
I feel like qmake just can't find the right .qmake.cache / .qmake.conf / .qmake.stash if QtAV.pro is included as a subproject

**Comment by nejatafshar** [comment]
@ElderOrb You are right. In order to use QtAV without installing to Qt directory, you should make some modifications. You can make the modules compile statically and change QmlAV modules to a regular static lib instead of a plugin. After that you should link to QtAV properly. For example if it's in your project as a subdir project, linking would be something like:

```
win32: {
    CONFIG(release, debug|release): {
        LIBS += -L$$OUT_PWD/../3rdparty/QtAV/lib_win_x86_64/ -lQtAV
        PRE_TARGETDEPS += $$OUT_PWD/../3rdparty/QtAV/lib_win_x86_64/QtAV.lib
    }
    else:CONFIG(debug, debug|release): {
        LIBS += -L$$OUT_PWD/../3rdparty/QtAV/lib_win_x86_64/ -lQtAVd
        PRE_TARGETDEPS += $$OUT_PWD/../3rdparty/QtAV/lib_win_x86_64/QtAVd.lib
    }
}
else:linux: {
    CONFIG(release, debug|release): {
        LIBS += -L$$OUT_PWD/../3rdparty/QtAV/lib_linux_x86_64/ -lQtAV
        PRE_TARGETDEPS += $$OUT_PWD/../3rdparty/QtAV/lib_linux_x86_64/libQtAV.a
    }
    else:CONFIG(debug, debug|release): {
        LIBS += -L$$OUT_PWD/../3rdparty/QtAV/lib_linux_x86_64/ -lQtAVd
        PRE_TARGETDEPS += $$OUT_PWD/../3rdparty/QtAV/lib_linux_x86_64/libQtAVd.a
    }
    LIBS += -ldl
}

INCLUDEPATH += $$PWD/../3rdparty/QtAV/src
DEPENDPATH += $$PWD/../3rdparty/QtAV/src
INCLUDEPATH += $$PWD/../3rdparty/QtAV/qml
DEPENDPATH += $$PWD/../3rdparty/QtAV/qml
```
If you use QmlAV module in your code, you should expose the qml types to qml engine before loading the main qml file like:

```
#include "QmlAV/QmlAVPlayer.h"
#include "QmlAV/QuickFBORenderer.h"
#include "QtAV/VideoCapture.h"
...
qmlRegisterType<QmlAVPlayer>("QtAV", 1, 7, "AVPlayer");
qmlRegisterType<QuickFBORenderer>("QtAV", 1, 7, "VideoOutput2");
qmlRegisterType<QtAV::VideoCapture>("QtAV", 1, 7, "VideoCapture");
```

I have tested this and it worked. I have also changed the project to be used directly independent of installing to Qt dir. One last thing is that I merged QmlAV module in to QtAV to have only one static lib.

**Comment by ElderOrb** [comment]
I used a bit different approach here: https://github.com/ElderOrb/qctools/tree/qtav but it allows me to build QtAV as a subtree project and moreover specify optional ffmpeg path for QtAV if required. Works on windows/macos/linux
